### ISSUE TITLE
PriceGapFill under CarriedPrice seeds a training window that opens inside a gap from the last training price, a look-ahead the online form shares

### ISSUE THREAD

**Issue by dcelisgarza** [dcelisgarza/PortfolioOptimisers.jl#1068]
Found by PR 625 review piece Q9 (map #955, ADR 0130). Measured at `467f40baa5`.

## The defect

`PriceGapFill` under `CarriedPrice` fits one value per asset, the **last observed training price**, and `apply_preprocessing` seeds every column's carry-forward with it (`src/03_InputData/08_PriceGapFill.jl`, `gap_fill_column!(::CarriedPrice, …)`: `carry = v` before the walk starts). ADR 0130 says the seed exists so that *"a window that opens inside a gap fills from the last observed training price"*, and for the **test** window that is right: the seed precedes every row it is written into.

For the **training** window itself it is a look-ahead. A `Pipeline` applies the fitted step to the training window before it fits the prior, and when that window opens inside an asset's suspension — the span is `true` there, the price is `NaN` — the seed written into the leading rows is a price from the **end** of the same window.

## Reproduction

```julia
using PortfolioOptimisers, TimeSeries, Dates
ts = Date(2020, 1, 1):Day(1):Date(2020, 1, 12)
P = [100.0 NaN 50.0 10.0; 101.0 NaN 51.0 10.1; 102.0 NaN 52.0 10.2; 103.0 20.0 53.0 10.3
     NaN 20.5 54.0 10.4; NaN 21.0 55.0 10.5; 106.0 21.5 56.0 10.6; 107.0 22.0 57.0 10.7
     108.0 22.5 58.0 10.8; 109.0 23.0 NaN 10.9; 110.0 23.5 NaN 11.0; 111.0 24.0 NaN 11.1]
pr = price_ingestion(PriceIngestion(), TimeArray(collect(ts), P, ["A", "B", "C", "D"]))
w = PortfolioOptimisers.port_opt_view(pr, 5:12)        # a train window opening inside A's suspension
res = fit_preprocessing(PriceGapFill(), w)
res.v[findfirst(==(:A), res.nx)]                        # 111.0 — the LAST price of the window
values(apply_preprocessing(res, w).X)[:, 1]'            # [111.0 111.0 106.0 107.0 108.0 109.0 110.0 111.0]
```

The raw column is `[NaN NaN 106 107 108 109 110 111]`. After the fill, the first return the conversion computes for `A` inside this window is `106 / 111 - 1 = -4.5 %`, a move the market never printed, computed with a price seven observations in the future. Under the map's own reading (ADR 0131, a return needs two consecutive prices) the two leading cells are a **Held Gap** and should stay non-finite on the training window; only a window that *follows* the fitted one has a last training price to carry in.

## The same shape in the online form (Q11's, #861)

`partial_fit_transform(est::PriceGapFill, pr)` in `src/03_InputData/18_PreprocessingPartialFit.jl` seeds a column with `v[j]` when the state holds a price, and otherwise with **the block's own last observed price** (`vals[t, j]` for `t = findlast(observed)`). On the first block that is the same look-ahead: a leading in-span gap of the block is filled from a price later in the block. The batch and the online forms should take the same rule.

## The fix I would make, and why it is not made here

The seed must apply only to observations **after** the fitted window. That needs the fitted object to know where the fitted window ended, so `PriceGapFillResult` gains a field — the last training timestamp, or the training end as the carrier's clock reads it — and `gap_fill_column!(::CarriedPrice, …)` starts the walk with **no** carry, takes the seed only once the walk passes that point without having met an observed price, and takes an observed price as the carry as it does now. On the training window every row is at or before the end, so the seed is never used and a leading in-span gap stays a Held Gap; on the test window every row is after it, so the seed is used exactly as ADR 0130 describes; on an overlapping window the rule interpolates.

The `Num_VecToScaM` arm is unaffected: its fitted value is a reduction over the training window, and replaying a fitted statistic on the window it was fitted on is the accepted meaning of fitted state.

It is not paid in Q9 because it changes a result type's fields, which reaches the online read-out `fit_preprocessing(est::PriceGapFill)` and `PriceGapFillState` in `18_PreprocessingPartialFit.jl` (map #861's file, under Q11's review), ADR 0130 § *Names* (the result "holds `nx` and one value per asset"), and the docstrings and `test_06g` that pin the seed. One session, both forms, one rule.

## Where it is stated

- `src/03_InputData/08_PriceGapFill.jl`: `gap_fill_column!(::CarriedPrice, …)`, `apply_preprocessing(::PriceGapFillResult, ::PricesResult)`, and the `PriceGapFill` docstring's *"The fitted value is a seed"* paragraph.
- `src/03_InputData/18_PreprocessingPartialFit.jl`: `partial_fit_transform(est::PriceGapFill, pr)`, the `seed = … vals[t, j]` arm.
- `docs/adr/0130-…md` § *Names*.
- `test/test_06g_price_gap_fill.jl`, `test/test_24f_pipeline_partial_fit.jl`.

**Comment by dcelisgarza** [comment]
Found by PR 625 review piece Q9, report issuecomment-5665667192 (finding 1); the paid findings of that piece are `d940aef074` on `dev`.

**Comment by dcelisgarza** [comment]
Fixed in c75f6282e3 on `dev`. `PriceGapFillResult` and `PriceGapFillState` record `te`, the training window's last timestamp; `gap_fill_column!(::CarriedPrice, …)` takes `t0` and writes the seed only from the first observation after it, so the training window's leading in-span gap stays a Held Gap and a later window is seeded as ADR 0130 describes. `partial_fit_transform(::PriceGapFill, …)` takes the same rule: a column the state has not priced has no seed. ADR 0130 § Names and ADR 0142 fact 1 rewritten in place (both drafts). Pinned in `test_06g` (the reproduction, the later window, the overlap, the reduction arm, the walk) and `test_24f` (online against batch on a history opened inside a suspension). No unit was added, so no sweep sub-issue was filed; `sweep_check` flagged `06_OnlinePipeline.jl` on a one-argument constructor update only.
